### ISSUE TITLE
using cloud.config.server.git.uri=file:/some/path causes spring-cloud-config-monitor to watch all files in .git/

### ISSUE THREAD

**Issue by ulisesbocchio** [spring-cloud/spring-cloud-config#350]
Using cloud.config.server.git.uri=file:/some/path watches the .git/ directory for changes. This results on a lot of refresh calls when properties are retrieved, since the repository always performs a pull that seems to change a bunch of files inside the .git/ directory.

My current fix consists in a custom FileMonitorConfiguration that overwrite the `preVisitDirectory` method of the anonymous file visitor in `FileMonitorConfiguration.walkDirectory` this way:

```java
                @Override
                public FileVisitResult preVisitDirectory(Path dir,
                                                         BasicFileAttributes attrs) throws IOException {
                    FileVisitResult fileVisitResult = super.preVisitDirectory(dir, attrs);
                    if(!dir.toUri().toString().contains(".git")) {
                        registerWatch(dir);
                        return fileVisitResult;
                    }
                    return FileVisitResult.SKIP_SUBTREE;
                }
```

It'll be nice to be able to customize the watcher providing either the file visitor, a spring abstraction of it, or  to honor the search-paths property.

**Comment by dsyer** [comment]
The ".git" hierarchy is already explicitly excluded from the change event set, but I can see that it is more efficient to exclude it from the watcher completely.

**Comment by dsyer** [comment]
If you have a proposal for a configuration or extension point, please open another issue and/or create a pull request. Thanks.

**Comment by ulisesbocchio** [comment]
Alright, thanks for fix. While that works, please consider adding `return FileVisitResult.SKIP_SUBTREE` when ".git" is detected to avoid even continue walking down the directory and effectively skip the whole ".git" subtree.
 
I was actually getting stuff from the .git directory on the change event set, not files, but directories, a bunch of them. This is the excerpt from FileMonitorConfiguration.filesFromEvents that's causing the issue:

```java
                  if (file.isDirectory()) {
                        files.addAll(walkDirectory(file.toPath()));
                    }
                    else {
                        if (!file.getPath().contains(".git") && !PatternMatchUtils
                            .simpleMatch(this.excludes, file.getName())) {
                            if (log.isDebugEnabled()) {
                                log.debug("Watch Event: " + event.kind() + ": " + file);
                            }
                            files.add(file);
                        }
                    }
```
basically the check for ".git" only happens if it's a file.
